zrml/court/src/lib.rs

```diff
@@ -1272,6 +1337,9 @@ mod pallet {
                             court_participant: who.clone(),
                             consumed_stake,
                             joined_at,
+                            last_join_at,
+                            pre_period_join_stake,
+                            pre_period_join_at,
```

Comment:
I am really having a hard time understanding why we need three different join dates. I think we could achieve the same with two and only storing `last_join_inflation_period_index` and `last_join_stake` additionally to existing fields. Then, whenever `pool_join` is called, 
```rust
if last_join_inflation_period_index != cur_idx { 
   last_join_inflation_period_index = cur_idx; 
   last_join_stake = amount - amount_previous ;
} else {
  last_join_stake += amount - amount_previous;
}
```

Now, whenever an inflation payout happens, payout is calculated as follows:
```rust
// cur_idx is the inflation period index for the period that is evaluated.
let payout = if last_join_inflation_period_index != cur_idx { 
  amount;
} else {
  amount - last_join_stake;
}

Reply:
First of all, let me explain my approach. It uses `pre_period_join_stake` and `pre_period_join_at` to store the relevant information about the stake that already has passed one inflation period, which is the requirement for inflation emission.

Assume there are three buckets A, B and C. Each bucket represents one inflation period. At the end of each inflation period, but before the next inflation period begins, the inflation distribution happens. Your first join happens in A (pre_period_stake = amount; pre_period_join_at = now). Inflation period A ends and no reward is distributed to you. Your second join happens in B (pre_period_stake unchanged; pre_period_join_at unchanged). At the end of B the inflation reward according to your _first_ join is distributed (using pre_period_* information). You don't join in C and the inflation period ends. Now you would expect to get inflation rewards based on your second join. But how would we know the current `stake` was actually already at least one inflation period old? In order to get `last_join_at`, we need to save it for each join (last_join_at is always reassigned to now at every join). If we didn't have `last_join_at`, we wouldn't know if one inflation period passed for the `pool_item.stake`. Here is another edge case: Say you would have joined in C. In this case you would overwrite pre_period_stake and pre_period_join_at by the information of the join in B. But for overwriting `pre_period_join_at` you need to have `last_join_at` and need to check if it was in a previous inflation period. If it was in a previous inflation period, then overwrite `pre_period_join_at`, otherwise use the old `pre_period_join_at`.

Now about your approach:

```rust
if last_join_inflation_period_index != cur_idx { 
  last_join_inflation_period_index = cur_idx;
  ...
else {
  ...
}
```

Ok, this would the last_join_inflation_period_index to be updated at every join.

```rust
if last_join_inflation_period_index != cur_idx { 
  last_join_inflation_period_index = cur_idx;
  last_join_stake = amount - amount_previous;
else {
  last_join_stake = last_join_stake + amount - amount_previous;
}
```

> last_join_stake = amount - amount_previous;

I assume `amount_previous` is the `pool_item.stake` of the last join? In that case `last_join_stake` would be only the **additional** stake since the last join. Why should the additional stake be relevant? We currently reward based on the full stake at a particular point in time.

> last_join_stake = last_join_stake + amount - amount_previous;

The above is done if last_join_inflation_period_index == cur_idx. I can't make sense of this formula. Why would you add the current join `amount` onto the last_join_stake?

```rust
// cur_idx is the inflation period index for the period that is evaluated.
let payout = if last_join_inflation_period_index != cur_idx { 
  amount;
} else {
  amount - last_join_stake;
}
```

I can follow the first case. There you want to use the current `pool_item.stake` for the inflation reward calculation, if the last join happened in a previous inflation period. That is fine.

I am confused about `amount - last_join_stake`, because it would only account for the additional stake since the last join.

In general the idea about storing inflation period indices instead of block numbers can be done.

Reply:
https://github.com/zeitgeistpm/zeitgeist/pull/1285/commits/ed69c2c9cfc13501458071938fcbb7c7479bae8a

Reply:
https://github.com/zeitgeistpm/zeitgeist/pull/1285/commits/8ef4484002ae88f04faaba5aa522fe6dd23478f4